classlib: fix SimpleNumber asTimeString for negative values

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation
Negative `SimpleNumber -asTimeString` returned incorrect string.
<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->
Previously:
```supercollider
-1.asTimeString;
-> 0-1:23:59:59.000
```
Now:
```supercollider
-1.asTimeString;
-> -00:00:01.000
```

## Types of changes

<!-- Delete lines that don't apply -->

- Bug fix

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -695,20 +695,24 @@ SimpleNumber : Number {
 		} {
 			days.asString.padLeft(3, "0").add($:);
 		};
+		if(isNegative, {days = "-" ++ days});
```
in general we use trailing closure for `if` in the class library. i've added a style point for that here: https://github.com/supercollider/supercollider/wiki/Code-style-guidelines#trailing-closure-syntax

can you change this one here please?

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -695,20 +695,24 @@ SimpleNumber : Number {
+		var number, decimal, days, hours, minutes, seconds, mseconds, isNegative = false;
+		if(this < 0, {
+			isNegative = true;
+		});
```
this can be simplified to `var number, decimal, days, hours, minutes, seconds, mseconds, isNegative = this < 0;`

brianlheim commented:
@dyfer gentle reminder to label your PRs!

dyfer commented:
> @dyfer gentle reminder to label your PRs!

Sorry I missed it! And thanks!

dyfer commented:
Thanks for a quick feedback!

I changed the if statement in the code that I added, but did not touch the one that was there already, should I change that too?

As for
> - `-0.0.asTimeString` 

I'm not sure what the proper output should be. Intuitively I'd expect
```supercollider
-0.0.asTimeString;
-> -00:00:00.000
```
but since we don't seem to be able to differentiate between `-0.0` and `0.0` (`-0.0 == 0.0 -> true`), I don't know how to achieve that. Currently this method gives
```supercollider
-0.0.asTimeString;
-> 00:00:00.000
```
In that case, should this be the expected result of the test?

dyfer commented:
...I guess I could check for identity:
```
-0.0 === 0.0
```
in the method, as a special case. Is this the way to go?

brianlheim commented:
> I changed the if statement in the code that I added, but did not touch the one that was there already, should I change that too?

that one seems right already, not sure what you're asking

> As for -0.0.asTimeString I'm not sure what the proper output should be

sorry, should have clarified. i think what it does now is fine, and actually makes more sense in my opinion as "-0" is not an intuitively sensible number for most people. this is also a lossy conversion so it's not vital to make sure everything round-trips between this and asSecs perfectly. i just want to pin that behavior so we can be aware if it changes in the future.

dyfer commented:
> that one seems right already, not sure what you're asking

My bad! I was in a rush and misread the code.

> sorry, should have clarified. i think what it does now is fine, and actually makes more sense in my opinion as "-0" is not an intuitively sensible number for most people. this is also a lossy conversion so it's not vital to make sure everything round-trips between this and asSecs perfectly. i just want to pin that behavior so we can be aware if it changes in the future.

Sorry, wanted to be done with this and followed that identity path. So now should I revert it and leave the behavior of 

```
-0.0.asTimeString;
-> 00:00:00.000
```
?

dyfer commented:
```supercollider
"-00:00:00.000".asSecs;
-> -0.0
```
so maybe that explicit check for `-0.0` should stay?

brianlheim commented:
took some time to think about this one.

i don't think we should output a timestamp saying `-0`, and the main reason is that while hardware makes an effort to ensure `-0` and `0` as floating point values compare equal, there is no such assistance for the strings `"-00:00:00"` and `"00:00:00"`. by outputting both we now put the burden on the programmer to handle this one case where two strings represent the same time stamp but do not compare equal.

furthermore, the knowledge that the input value that produced `"-00:00:00"` was `-0` or rounded up from slightly below 0 is, in my view, of marginal utility at best, and distracting or confusing at worst.

another reason i prefer to not have a `-0` is that fixed-point decimals such as these timestamps are conceptually closer to integers than floating point numbers. importantly, the interval between successive values remains fixed (like integers) rather than increasing exponentially (as with floats).

finally, in my experience the vast number of numerical comparisons are typically done with `==`, for good reason -- `0` and `-0` wouldn't compare equal under identity `===`, but they do under equality `==`. and, `asTimeString` followed by `asSecs` always produces a float regardless of whether the original number was an integer. for that reason, i think ensuring that a round trip between `asTimeString` and `asSecs` preserves identity cannot be a goal here.

so, i think this PR should be modified such that `-0.0`, `-0.0001`, etc. all produce the result `00:00:00[.000]` with no negative sign. that means checking for negative values after rounding.

dyfer commented:
Thanks @brianlheim for you thoughts. I changed it so it 
```supercollider
-0.0.asTimeString;
-> 00:00:00.000
```

> that means checking for negative values after rounding.

AFAICT I only need to not check for identity, since 
```supercollider
-0.0 < 0;
-> false
```
I'm just pointing it out to make sure I'm not doing something wrong.

brianlheim commented:
> AFAICT I only need to not check for identity,

you also need to handle inputs like -.001 which are negative but may round up to 0 for the requested precision. please make sure to add a test for that.

dyfer commented:
> you also need to handle inputs like -.001 which are negative but may round up to 0 for the requested precision. please make sure to add a test for that.

done!
